subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/asm/ClassDependenciesVisitor.java

```diff
@@ -49,39 +49,24 @@
     private final RetentionPolicyVisitor retentionPolicyVisitor;
     private final AnnotationVisitor annotationVisitor;
 
-    private ClassDependenciesVisitor(IntSet constantsCollector) {
-        this(constantsCollector, null, null, null, null);
-    }
-
-    private ClassDependenciesVisitor(IntSet constantsCollector, Set<String> types, Predicate<String> typeFilter, ClassReader reader, StringInterner interner) {
+    private ClassDependenciesVisitor(Predicate<String> typeFilter, ClassReader reader, StringInterner interner) {
         super(API);
-        this.constants = constantsCollector;
-        this.types = types;
-        this.superTypes = types == null ? null : Sets.<String>newHashSet();
-        this.methodVisitor = types == null ? null : new MethodVisitor();
-        this.fieldVisitor = types == null ? null : new FieldVisitor();
-        this.retentionPolicyVisitor = types == null ? null : new RetentionPolicyVisitor();
-        this.annotationVisitor = types == null ? null : new AnnotationVisitor();
+        this.constants = new IntOpenHashSet(2);
+        this.types = Sets.newHashSet();
+        this.superTypes = Sets.newHashSet();
+        this.methodVisitor = new MethodVisitor();
+        this.fieldVisitor = new FieldVisitor();
+        this.retentionPolicyVisitor = new RetentionPolicyVisitor();
+        this.annotationVisitor = new AnnotationVisitor();
         this.typeFilter = typeFilter;
-        if (reader != null) {
-            collectClassDependencies(reader);
-        }
         this.interner = interner;
+        collectClassDependencies(reader);
```

Comment:
This was a bug in the previous implementation - All types collected by this method were not interned. This can no longer happen, as the interner is no longer nullable.